src/permutations.rs

```diff
@@ -222,22 +184,34 @@ impl CompleteState {
             }
         }
     }
+}
 
-    /// Returns the count of remaining permutations, or None if it would overflow.
-    fn remaining(&self) -> Option<usize> {
-        match self {
-            &CompleteState::Start { n, k } => {
-                if n < k {
-                    return Some(0);
-                }
-                (n - k + 1..=n).try_fold(1usize, |acc, i| acc.checked_mul(i))
+impl PermutationState {
+    fn size_hint_for(&self, n: usize) -> SizeHint {
+        // At the beginning, there are `n!/(n-k)!` items to come.
+        let at_start = |n, k| {
+            debug_assert!(n >= k);
```

Comment:
I wrote that in my previous branch some weeks ago so I was not much familiar with it, and take a glance at is definitely not enough as it requires a bit of thinking.

This `debug_assert!` only occurs with `Start` and `Buffered` variants.
At definition, we prefill the lazy buffer with `k` values. It has enough values (or we would have the `End` variant) so `vals.len() >= k` (`vals.len()==k` at definition, more later).
`size_hint_for` is then called with:
- in the case of `count`: `n = vals.count() >= vals.len()` (see lazy buffer for `>=`) ;
- in the case of `size_hint`: `n = vals.size_hint().0 >= vals.len()` (see lazy buffer).
Similar for `vals.size_hint().1`.

So in each case: `n >= vals.len() >= k`. Basically, it holds because we prefilled with `k` values.

Reply:
However, I'm considering to soon work on making all our iterators lazy (such as #602) and I'll surely turn that assertion into `if n < k { return (0, Some(0)); }` (and move the "prefill the lazy buffer" part).